### PR TITLE
Improved kontena-ipam-cleanup

### LINKED ISSUE #53: kontena-ipam-cleanup is not safe against concurrent IpamDriver.RequestAddress
* Only delete etcd nodes older than e.g. 5 minutes during cleanup, to avoid conflicts with ongoing allocations.

### PR DESCRIPTION
**Pull request by SpComb**
Follows #62

Replaces #60

Fixes #53 

* Split out `DockerClient` with `networks_addresses` having `yield pool, address`
* Refactor `Addresses::Cleanup < Mutations::Command` to be per-pool
* Split out `Commands::Cleanup` with options for Docker networks cleanup
* Track the etcd-index for safe address cleanup

    * Get the current etcd-index
    * Sleep for `--quiesce-sleep=SECONDS` to allow pending create operations to complete
    * Scan allocated addresses from Docker
    * Delete etcd addresses with `modified-index < etcd-index` that are not in use

* Add `/KontenaIPAM.Cleanup` API

## `/KontenaIPAM.Cleanup`

Used for non-Docker based cleanup with externally managed container addresses.

#### Response
```
{
    "EtcdIndex": X
}
``` 

#### Request
```
{
    "EtcdIndex": X,
    "PoolID": "kontena",
    "Addresses": [ '10.81.1.A' ]
}
```
   

## `bin/kontena-ipam-cleanup --docker-networks`

Used for Docker API based cleanup operations.

Executed as a separate process to not block the Kontena IPAM API.

```
Usage: kontena-ipam-cleanup [--docker-networks]
        --quiesce-sleep=SECONDS      Wait for concurrent operations to complete
        --pool=POOL                  Scan pool for cleanup even if not in use by any containers
        --docker-networks            Scan Docker network endpoints
```

Without any options, the script does nothing.

### CODE REVIEW

**Comment by SpComb** [comment]
Updated to automatically handle the `--etcd-index-upto=` for `--docker-networks`.

Testing with some injected sleep, launching a new Docker container after scanning the network but before the cleanup mutation:

```
I, [2016-11-11T16:24:09.463736 #10]  INFO -- EtcdClient: Connected to http://localhost:2379 version {"etcdserver"=>"2.3.3", "etcdcluster"=>"2.3.0"} 
D, [2016-11-11T16:24:09.466016 #10] DEBUG -- EtcdClient: get / {}: directory@19789: docker/ kontena/ coreos.com/
I, [2016-11-11T16:24:09.466969 #10]  INFO -- Commands::Cleanup: Cleanup Docker networks...
D, [2016-11-11T16:24:09.467119 #10] DEBUG -- Commands::Cleanup: Cleanup upto etcd_index=19789
D, [2016-11-11T16:24:09.469783 #10] DEBUG -- DockerClient: Scanning network: none
D, [2016-11-11T16:24:09.470456 #10] DEBUG -- DockerClient: Skip non-kontena network: IPAM.Driver=default
D, [2016-11-11T16:24:09.471719 #10] DEBUG -- DockerClient: Scanning network: docker_gwbridge
D, [2016-11-11T16:24:09.472349 #10] DEBUG -- DockerClient: Skip non-kontena network: IPAM.Driver=default
D, [2016-11-11T16:24:09.473582 #10] DEBUG -- DockerClient: Scanning network: kontena
D, [2016-11-11T16:24:09.474112 #10] DEBUG -- DockerClient: Kontena IPAM network kontena: pool=kontena containers=0
I, [2016-11-11T16:24:09.474311 #10]  INFO -- Commands::Cleanup: Cleanup Docker network kontena using Kontena IPAM pool kontena with 0 local Docker container endpoints
D, [2016-11-11T16:24:09.474405 #10] DEBUG -- Commands::Cleanup: Sleeping for 10s...
D, [2016-11-11T16:24:19.474901 #10] DEBUG -- Commands::Cleanup: Continuing
D, [2016-11-11T16:24:19.490081 #10] DEBUG -- EtcdClient: get /kontena/ipam/pools/kontena {}: node@19796: {"subnet":"10.81.0.0/16","iprange":"10.81.128.0/17","gateway":"10.81.0.1/16"}
D, [2016-11-11T16:24:19.494792 #10] DEBUG -- Addresses::Cleanup: checking pool kontena for node=coreos-dev-01 with active_addresses=0
D, [2016-11-11T16:24:19.506023 #10] DEBUG -- EtcdClient: get /kontena/ipam/addresses/kontena/ {}: directory@19796: 10.81.0.1 10.81.128.0 10.81.128.1
D, [2016-11-11T16:24:19.508199 #10] DEBUG -- Addresses::Cleanup: checking address 10.81.0.1: #<Address:0x0055a4b00f2828 @etcd_node=#<Etcd::Node:0x0055a4b00f2dc8 @created_index=1103, @modified_index=1103, @ttl=nil, @key="/kontena/ipam/addresses/kontena/10.81.0.1", @value="{\"address\":\"10.81.0.1/16\",\"node\":\"coreos-dev-01\"}", @expiration=nil, @dir=nil>, @pool="kontena", @id="10.81.0.1", @address=#<IPAddr: IPv4:10.81.0.1/255.255.0.0>, @node="coreos-dev-01">
D, [2016-11-11T16:24:19.513686 #10] DEBUG -- Addresses::Cleanup: ...gateway, skipping
D, [2016-11-11T16:24:19.513686 #10] DEBUG -- Addresses::Cleanup: checking address 10.81.128.0: #<Address:0x0055a4b00f1cc0 @etcd_node=#<Etcd::Node:0x0055a4b00f2da0 @created_index=7735, @modified_index=7735, @ttl=nil, @key="/kontena/ipam/addresses/kontena/10.81.128.0", @value="{\"address\":\"10.81.128.0/16\",\"node\":\"kontena-ipam\"}", @expiration=nil, @dir=nil>, @pool="kontena", @id="10.81.128.0", @address=#<IPAddr: IPv4:10.81.128.0/255.255.0.0>, @node="kontena-ipam">
D, [2016-11-11T16:24:19.513686 #10] DEBUG -- Addresses::Cleanup: ...not managed by this node, skipping
D, [2016-11-11T16:24:19.513686 #10] DEBUG -- Addresses::Cleanup: checking address 10.81.128.1: #<Address:0x0055a4b00f0f28 @etcd_node=#<Etcd::Node:0x0055a4b00f2d78 @created_index=19791, @modified_index=19791, @ttl=nil, @key="/kontena/ipam/addresses/kontena/10.81.128.1", @value="{\"address\":\"10.81.128.1/16\",\"node\":\"coreos-dev-01\"}", @expiration=nil, @dir=nil>, @pool="kontena", @id="10.81.128.1", @address=#<IPAddr: IPv4:10.81.128.1/255.255.0.0>, @node="coreos-dev-01">
D, [2016-11-11T16:24:19.515220 #10] DEBUG -- Addresses::Cleanup: ...recently allocated, skipping
```

File: app/commands/cleanup.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    attr_accessor :docker_networks
+    attr_accessor :pool
+    attr_accessor :addresses
+
+    def self.parse!
+      command = new
+
+      OptionParser.new do |opts|
+        opts.banner = "Usage: kontena-ipam-cleanup [--docker-networks] [--pool POOL IPADDR...]"
+        opts.on("--etcd-index-upto=INDEX", Integer, "Only cleanup nodes created <= given etcd index") do |index|
```
Comment (jnummelin) [inline app/commands/cleanup.rb]: Some sane default would be really nice for this, e.g. `current_index - 10`. That way agent, and any other possible user, would not really have to understand ETCD indexes at all.

File: lib/etcd/fake_server.rb [hunk 2]
```diff
@@ -126,18 +140,26 @@ def read(key)
...
+    node
+  end
+
+  def update(node, value)
+    @index += 1
+
+    node.modify(@index, value)
   end
 
   def remove(node, toplevel: true)
```
Comment (jnummelin) [inline lib/etcd/fake_server.rb]: `remove` should probably also increment the index

File: app/commands/cleanup.rb (hunk 1 above)
Comment (SpComb) [inline app/commands/cleanup.rb]: Not sure there is any sane fixed `current_index - 10` offset, other things can also use the same etcd server, so we can't predict how fast the indexes are going to rise.

File: lib/etcd/fake_server.rb (hunk 2 above)
Comment (SpComb) [inline lib/etcd/fake_server.rb]: Done

**Comment by SpComb** [comment]
Changed to split out the `EtcdModel` changes for etcd-index support into #62, and replace `--etcd-index-upto=` with `--docker-container-*-label=...`.

**Comment by SpComb** [comment]
Updated to split `{GET,POST} /KontenaIPAM.Cleanup` and `bin/kontena-ipam-cleanup --docker-networks`

**Comment by SpComb** [comment]
Rebased over #62 against master.

File: app/mutations/addresses/cleanup.rb [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+      string :pool_id
+      array :addresses do
+        ipaddr
+      end
+    end
+
+    optional do
+      integer :etcd_index_upto
+    end
+
```
Comment (jnummelin) [inline app/mutations/addresses/cleanup.rb]: Is etcd index really optional? A bit unsafe cleanup without it.

Comment (SpComb) [inline app/mutations/addresses/cleanup.rb]: Changed to required.

**Pull request merged**